src/lib/core/CHIPError.h

```diff
@@ -91,92 +94,248 @@ class ChipError
-    static constexpr bool IsRange(Range range, ErrorType error)
+    /**
+     * Test whether @a error belongs to the Range @a range.
+     */
+    constexpr bool IsRange(Range range) const
     {
-        return (AsInteger(error) & MakeMask(kRangeStart, kRangeLength)) == MakeField(kRangeStart, static_cast<StorageType>(range));
+        return (mError & MakeMask(kRangeStart, kRangeLength)) == MakeField(kRangeStart, static_cast<StorageType>(range));
     }
 
-    /// Get the `Range` to which the @a error belongs.
-    static constexpr Range GetRange(ErrorType error)
+    // DO NOT USE. This is transitional and will be removed soon. Instead of `IsRange(status)` use `status.IsRange()`.
+    static constexpr bool IsRange(Range range, ChipError error)
+    {
+        return (error.mError & MakeMask(kRangeStart, kRangeLength)) == MakeField(kRangeStart, static_cast<StorageType>(range));
+    }
+
+    /**
+     * Get the Range to which the @a error belongs.
+     */
+    constexpr Range GetRange() const { return static_cast<Range>(GetField(kRangeStart, kRangeLength, mError)); }
+
+    // DO NOT USE. This is transitional and will be removed soon. Instead of `GetRange(status)` use `status.GetRange()`.
+    static constexpr Range GetRange(ChipError error)
     {
-        return static_cast<Range>(GetField(kRangeStart, kRangeLength, AsInteger(error)));
+        return static_cast<Range>(GetField(kRangeStart, kRangeLength, error.mError));
     }
 
-    /// Get the encapsulated value of an @a error.
-    static constexpr ValueType GetValue(ErrorType error) { return GetField(kValueStart, kValueLength, AsInteger(error)); }
+    /**
+     * Get the encapsulated value of an @a error.
+     */
+    constexpr ValueType GetValue() const { return GetField(kValueStart, kValueLength, mError); }
 
-    /// Test whether if @a value can be losslessly encapsulated in a `CHIP_ERROR`.
-    static constexpr bool CanEncapsulate(ValueType value) { return FitsInField(kValueLength, value); }
+    // DO NOT USE. This is transitional and will be removed soon. Instead of `GetValue(status)` use `status.GetValue()`.
+    static constexpr ValueType GetValue(ChipError error) { return GetField(kValueStart, kValueLength, error.mError); }
 
-    /// Construct a `CHIP_ERROR` encapsulating @a value inside the @a range.
-    static constexpr ErrorType Encapsulate(Range range, ValueType value)
+    /**
+     * Test whether type @a T can always be losslessly encapsulated in a CHIP_ERROR.
+     */
+    template <typename T>
+    static constexpr bool CanEncapsulate()
+    {
+        return std::numeric_limits<typename std::make_unsigned<T>::type>::digits <= kValueLength;
+    }
+
+    /**
+     * Test whether if @a value can be losslessly encapsulated in a CHIP_ERROR.
+     */
+    template <typename T>
+    static constexpr bool CanEncapsulate(T value)
+    {
+        return CanEncapsulate<T>() || FitsInField(kValueLength, static_cast<ValueType>(value));
```

Comment:
Should we also check if `value` can be cast to `ValueType`?

Reply:
I _think_ we get that from `CanEncapsulate<T>()`, unless I'm missing some weird C++ corner, assuming `kValueLength` bits fit in `ValueType`. Perhaps I should add a bunch of static_asserts for the bit field constants.

Reply:
Added field assertions.